node/core/approval-voting/src/lib.rs

```diff
@@ -638,69 +637,54 @@ impl CurrentlyCheckingSet {
 	}
 }
 
+// Wraps `RuntimeInfo` and some metadata. On each new leaf `SessionInfo` is
+// cached. `RuntimeInfo` keeps the last `DISPUTE_WINDOW` number of sessions.
+struct SessionInfoProvider {
+	// `RuntimeInfo` caches sessions internally.
+	runtime_info: RuntimeInfo,
+	// Will be true if an error had occurred during the last session caching attempt
+	gaps_in_cache: bool,
+	// Highest session index seen so far. Also used to calculate the earliest one.
+	highest_session_seen: SessionIndex,
+}
+
+impl SessionInfoProvider {
+	fn earliest_session(&self) -> SessionIndex {
+		self.highest_session_seen.saturating_sub(DISPUTE_WINDOW.get() - 1)
```

Comment:
Again, I'm not sure why dispute window from the highest seen session is used to bail with `BlockFromAncientSession` for approval-voting. We should allow importing and producing messages related to down to the last finalized block. Or do you set `highest_session_seen` depending on each `head`?

Reply:
I don't set `highest_session_seen` for each head. This is wrong - earliest session should be the one of the last finalized block, as you already mentioned.

Reply:
This is also fixed. `earliest_session` is removed and the last finalized block is used instead.